File: Python/MaximumPathQualityOfAGraph.py
from typing import List


class Solution:
    def maximalPathQuality(self, values: List[int], edges: List[List[int]], maxTime: int) -> int:
        # Exhaustive backtracking. maxTime <= 100 and every edge costs >= 10, so a path has
        # at most 10 edges; with degree <= 4 the search tree is <= 4^10, pruned by remaining
        # time. visited[v] counts active visits so each node's value is added exactly once.
        n = len(values)
        adj: List[List[tuple[int, int]]] = [[] for _ in range(n)]
        for u, v, t in edges:
            adj[u].append((v, t))
            adj[v].append((u, t))

        visited = [0] * n
        best = 0

        def dfs(node: int, time_left: int, quality: int) -> None:
            nonlocal best
            if node == 0:
                best = max(best, quality)
            for v, t in adj[node]:
                if t <= time_left:
                    gain = values[v] if visited[v] == 0 else 0
                    visited[v] += 1
                    dfs(v, time_left - t, quality + gain)
                    visited[v] -= 1

        visited[0] = 1
        dfs(0, maxTime, values[0])
        return best


if __name__ == '__main__':
    s = Solution()
    print(s.maximalPathQuality([0, 32, 10, 43], [[0, 1, 10], [1, 2, 15], [0, 3, 10]], 49))  # 75
    print(s.maximalPathQuality([5, 10, 15, 20], [[0, 1, 10], [1, 2, 10], [0, 3, 10]], 30))  # 25
    print(s.maximalPathQuality([1, 2, 3, 4],
                               [[0, 1, 10], [1, 2, 11], [2, 3, 12], [1, 3, 13]], 50))        # 7
    print(s.maximalPathQuality([5], [], 100))                                                # 5

File: Go/MaximumPathQualityOfAGraph.go
//go:build ignore

package main

import "fmt"

type edge struct{ to, cost int }

func maximalPathQuality(values []int, edges [][]int, maxTime int) int {
	n := len(values)
	adj := make([][]edge, n)
	for _, e := range edges {
		u, v, t := e[0], e[1], e[2]
		adj[u] = append(adj[u], edge{v, t})
		adj[v] = append(adj[v], edge{u, t})
	}

	visited := make([]int, n)
	best := 0

	var dfs func(node, timeLeft, quality int)
	dfs = func(node, timeLeft, quality int) {
		if node == 0 && quality > best {
			best = quality
		}
		for _, e := range adj[node] {
			if e.cost <= timeLeft {
				gain := 0
				if visited[e.to] == 0 {
					gain = values[e.to]
				}
				visited[e.to]++
				dfs(e.to, timeLeft-e.cost, quality+gain)
				visited[e.to]--
			}
		}
	}

	visited[0] = 1
	dfs(0, maxTime, values[0])
	return best
}

func main() {
	fmt.Println(maximalPathQuality([]int{0, 32, 10, 43}, [][]int{{0, 1, 10}, {1, 2, 15}, {0, 3, 10}}, 49)) // 75
	fmt.Println(maximalPathQuality([]int{5, 10, 15, 20}, [][]int{{0, 1, 10}, {1, 2, 10}, {0, 3, 10}}, 30)) // 25
	fmt.Println(maximalPathQuality([]int{1, 2, 3, 4}, [][]int{{0, 1, 10}, {1, 2, 11}, {2, 3, 12}, {1, 3, 13}}, 50)) // 7
	fmt.Println(maximalPathQuality([]int{5}, [][]int{}, 100)) // 5
}
